File: solutions/go/protein-translation/protein_translation.go
package protein

import (
	"errors"
)

var ErrInvalidBase = errors.New("invalid RNA")
var ErrStop = errors.New("known stop codon")

func FromRNA(rna string) ([]string, error) {
	var result []string
	for i := 0; i < len(rna); i += 3 {
		protein, err := FromCodon(rna[i : i+3])
		if err == ErrStop {
			return result, nil
		}
		if err != nil {
			return nil, err
		}
		result = append(result, protein)
	}
	return result, nil
}

func FromCodon(codon string) (string, error) {
	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}

File: solutions/fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let proteins (rna: string) : string list =
    let fromCodon (strand: string) : (string * string) option =
        match strand[0..2] with
        | "AUG" -> Some("Methionine", strand[3..])
        | "UUU"
        | "UUC" -> Some("Phenylalanine", strand[3..])
        | "UUA"
        | "UUG" -> Some("Leucine", strand[3..])
        | "UCU"
        | "UCC"
        | "UCA"
        | "UCG" -> Some("Serine", strand[3..])
        | "UAU"
        | "UAC" -> Some("Tyrosine", strand[3..])
        | "UGU"
        | "UGC" -> Some("Cysteine", strand[3..])
        | "UGG" -> Some("Tryptophan", strand[3..])
        | "UAA"
        | "UAG"
        | "UGA"
        | "" -> None
        | _ -> failwith "invalid codon"

    List.unfold fromCodon rna
